### ISSUE TITLE
Verbose_ostream and std::wcerr

### ISSUE THREAD

**Issue by aothms** [CGAL/cgal#4249]
I imagine this is the curse of being a Windows user. At some point in my application I decided to use std::wcerr and std::wcout to have better Unicode output on Windows (at least I was made to believe so). The issue is that using std::cerr and std::wcerr interchangeably causes hard crashes. 

In my case I'm converting from Nef to Polyhedron, which hardcodes the verbosity to true [0], and uses a Verbose_ostream constructed on the stack with the default stream parameter [1] set to std::cerr [2].

[0] https://github.com/CGAL/cgal/blob/master/Nef_3/include/CGAL/Nef_polyhedron_3.h#L876
[1] https://github.com/CGAL/cgal/blob/master/Polyhedron/include/CGAL/Polyhedron_incremental_builder_3.h#L513
[2] https://github.com/CGAL/cgal/blob/master/Stream_support/include/CGAL/IO/Verbose_ostream.h#L39

In other words in my limited understanding I have no influence over this function writing to std::cerr.

Would you be willing to consider something like a preprocessor directive to govern this?

~~~
#ifndef VERBOSE_OSTREAM_OUT
#define VERBOSE_OSTREAM_OUT std::cerr
#endif

...

    Verbose_ostream( bool active = false, std::ostream& out = VERBOSE_OSTREAM_OUT)
        : b(active), o(&out){}
~~~

Maybe even a flag to force active to false?

~~~
#ifdef VERBOSE_OSTREAM_FORCE_INACTIVE
#define VERBOSE_OSTREAM_FORCE_INACTIVE_VALUE false
#else
#define VERBOSE_OSTREAM_FORCE_INACTIVE_VALUE true
#endif

...

    Verbose_ostream( bool active = false, std::ostream& out = VERBOSE_OSTREAM_OUT)
        : b(VERBOSE_OSTREAM_FORCE_INACTIVE_VALUE && active), o(&out){}
~~~

Thougts?

**Comment by sloriot** [comment]
Not related to the issue but you might not know that there are other Boolean operations in CGAL in the [Polygon Mesh Processing package](https://doc.cgal.org/latest/Polygon_mesh_processing/#title12). They are more limited than Nef_3 but way faster and still robust.

**Comment by lrineau** [comment]
You are right. That is a value issue.

And the situation might be even worst: all CGAL headers hardcode `std::istream` and `std::ostream`, instead of using the template `std::basic_istream` and `std::basic_ostream`. There are occurrences in 459 CGAL headers.

So the first suggestion about `#define VERBOSE_OSTREAM_OUT std::cerr` will not work, because nothing in CGAL has `operator<<` defined for `std::wostream`.

The second suggestion, to disable `Verbose_ostream`, will work, though.

**Comment by afabri** [comment]
So why not globally replacing `std::istream` then?

**Comment by lrineau** [comment]
As I say, there are occurrences in 459 CGAL headers, and it cannot be a simple search-and-replace, because `std::istream` is a class and `std::basic_istream` is a template (same for the `ostream` equivalent). And thus, for example, a declaration like:
```c++
template < class R >
std::ostream&
operator<<(std::ostream& os, const Point_2<R>& p);
```
would have to be turned into:
```c++
template < class R, typename CharT, typename Traits >
std::basic_ostream<CharT¸ Traits>&
operator<<(std::basic_ostream<CharT, Traits>& os, const Point_2<R>& p);
```

One can automatically replace `ostream` by `basic_ostream<CharT, Traits>`, but the addition of the two template parameters `CharT` and `Traits` is more difficult to deal with.

**Comment by afabri** [comment]
I would still call that globally replacing.   I guess there is an additional problem with wide strings then?

**Comment by lrineau** [comment]
[**@afabri**](https://github.com/afabri) commented on [Sep 25, 2019, 10:41 AM GMT+2](https://github.com/CGAL/cgal/issues/4249#issuecomment-534917733 "2019-09-25T08:41:43Z - Replied by Github Reply Comments"):
> I would still call that globally replacing.   I guess there is an additional problem with wide strings then?

`<CGAL/IO.h>` will have to be adapted, because it currently assumed that the type of character of a stream is `char` whereas is should be `typename Traits::char_type`, and same for `int` instead of `typename Traits::int_type`.

**Comment by aothms** [comment]
Thanks all for the insightful replies.

In the spirit of ugly hacks I came up with the following for my own purpose at least:

~~~
class wcerr_adapter : public std::ostringstream {
public:
	virtual ~stream_adapter() {
		auto s = str();
		std::wcerr << s.c_str() << std::flush;
	}
};

class adapted_wcerr {
public:
	static wcerr_adapter& get() {
		static wcerr_adapter s;
		return s;
	}
};

class Verbose_ostream {
    bool          b;
    std::ostream* o;
public:
    Verbose_ostream( bool active = false, std::ostream& out = adapted_wcerr::get())
~~~

Invent a class that inherits from ostringstream (i.e. ostream) so that no function prototypes have to be adapted. Wrap it in another class to get a static reference. Buffer everything and then write to wcerr (i.e. wide) on deletion / program termination (due to the static local variable).

It can be made a bit more general and less ugly probably. I can imagine people would wont to log to their own in memory stream or a file stream. This opens ways for end users to control this at runtime and not through the preprocessor.

If this is a path you would want to consider I can see if I can come up with something workable.
